### ISSUE TITLE
Writing DG vectors (dim >= 3) to XDMF gives node-centered fields

### ISSUE THREAD

**Issue by tianyikillua** [FEniCS/dolfinx#350]
Correct behavior when writing a **scalar** DG field to a XDMF file

``` xml
...
<Attribute Name="Test" AttributeType="Scalar" Center="Cell">
    <DataItem Dimensions="8 1" Format="HDF">u.h5:/VisualisationVector/0</DataItem>
</Attribute>
...
```

![scalar](https://user-images.githubusercontent.com/4027283/54682911-8259e480-4b10-11e9-9f13-224f0b2e8258.png)

However writing a **vector** DG field returns an **interpolated** nodal field

``` xml
...
<Attribute Name="Test" AttributeType="Vector" Center="Node">
    <DataItem Dimensions="8 3" Format="HDF">u.h5:/VisualisationVector/0</DataItem>
</Attribute>
...
```

![vector](https://user-images.githubusercontent.com/4027283/54682917-86860200-4b10-11e9-9ca7-eadeb26348ec.png)

MWE à la `dolfinx`
``` python
from dolfin import *
from dolfin.io import XDMFFile
import numpy as np

mesh = RectangleMesh(
    MPI.comm_world,
    [Point(0, 0)._cpp_object, Point(1, 1)._cpp_object], [2, 2],
    CellType.Type.triangle, dolfin.cpp.mesh.GhostMode.none)
cmap = dolfin.fem.create_coordinate_map(mesh.ufl_domain())
mesh.geometry.coord_mapping = cmap
num_cells = 8

# Write a scalar DG field to a XDMF file
# Correct behavior, Center="Cell" in XDMF
V = FunctionSpace(mesh, ("DG", 0))
u = Function(V, name="Test")
values = np.arange(num_cells)
u.vector()[:] = values.flatten()

with XDMFFile(MPI.comm_world, "u.xdmf") as f:
    f.write(u, 0)

# Write a vector (dim >= 3) DG field to a XDMF file
# With dim = 2, correct behavior, Center="Cell" in XDMF
# With dim >= 3, wrong behavior, Center="Node" in XDMF
dim = 3
V = VectorFunctionSpace(mesh, ("DG", 0), dim=dim)
v = Function(V, name="Test")
values = np.zeros((num_cells, dim))
values[:, 0] = np.arange(num_cells)
values[:, 1] = -np.arange(num_cells)

v.vector()[:] = values.flatten()

with XDMFFile(MPI.comm_world, "v.xdmf") as f:
    f.write(v, 0)
```

Suspect
``` c++
bool XDMFFile::has_cell_centred_data(const Function& u)
{
  // Test for cell-centred data
  std::size_t cell_based_dim = 1;
  for (std::size_t i = 0; i < u.value_rank(); i++)
    cell_based_dim *= u.function_space()->mesh()->topology().dim();
  return (u.function_space()->dofmap()->max_element_dofs() == cell_based_dim);
}
```

**Comment by garth-wells** [comment]
@jorgensd Is this still an issue, or has it been fixed?

**Comment by tianyikillua** [comment]
Still an issue.

``` python
from dolfin import *
from dolfin.cpp.mesh import CellType
from dolfin.io import XDMFFile
import numpy as np

mesh = RectangleMesh(
    MPI.comm_world,
    [np.array([0, 0, 0]), np.array([1, 1, 0])],
    [2, 2],
    CellType.triangle,
    dolfin.cpp.mesh.GhostMode.none,
)
cmap = dolfin.fem.create_coordinate_map(mesh.ufl_domain())
mesh.geometry.coord_mapping = cmap

# Write a vector (dim >= 3) DG field to a XDMF file
# With dim = 2, correct behavior, Center="Cell" in XDMF
# With dim >= 3, wrong behavior, Center="Node" in XDMF
dim = 3
V = VectorFunctionSpace(mesh, ("DG", 0), dim=dim)
v = Function(V, name="Test")

with XDMFFile(MPI.comm_world, "w.xdmf") as f:
    f.write(v, 0)
```

**Comment by jorgensd** [comment]
> @jorgensd Is this still an issue, or has it been fixed?

Do you want me to have a go at this? I havent really considered fixes all the XDMF stuff, as its not usable for higher order geometries.

**Comment by jorgensd** [comment]
> Still an issue.
> 
> ```python
> from dolfin import *
> from dolfin.cpp.mesh import CellType
> from dolfin.io import XDMFFile
> import numpy as np
> 
> mesh = RectangleMesh(
>     MPI.comm_world,
>     [np.array([0, 0, 0]), np.array([1, 1, 0])],
>     [2, 2],
>     CellType.triangle,
>     dolfin.cpp.mesh.GhostMode.none,
> )
> cmap = dolfin.fem.create_coordinate_map(mesh.ufl_domain())
> mesh.geometry.coord_mapping = cmap
> 
> # Write a vector (dim >= 3) DG field to a XDMF file
> # With dim = 2, correct behavior, Center="Cell" in XDMF
> # With dim >= 3, wrong behavior, Center="Node" in XDMF
> dim = 3
> V = VectorFunctionSpace(mesh, ("DG", 0), dim=dim)
> v = Function(V, name="Test")
> 
> with XDMFFile(MPI.comm_world, "w.xdmf") as f:
>     f.write(v, 0)
> ```
The built-in 2D meshes are 2D, even if you send in 3D coordinates, So dim=3 doesnt really makes sense in that setting. However, it is an issue with actual 3D meshes as well:
 ```
from dolfin import *
from dolfin.cpp.mesh import CellType, GhostMode
from dolfin.io import XDMFFile
import numpy as np
points = np.array([[-1., 0.5, 0], [0., 0., 0], [0., 1., 0], [1., 0.5, 0]])
cells = np.array([[0, 1, 2], [3, 1, 2]])
mesh = Mesh(MPI.comm_world, CellType.triangle, points, cells,
                    [], GhostMode.none)
cmap = dolfin.fem.create_coordinate_map(mesh.ufl_domain())
mesh.geometry.coord_mapping = cmap

dim = 3
V = VectorFunctionSpace(mesh, ("DG", 0), dim=dim)
v = Function(V, name="Test")
v.vector.array[0] = 1
with XDMFFile(MPI.comm_world, "w.xdmf") as f:
    f.write(v, 0)

```

**Comment by jorgensd** [comment]
@tianyikillua you can use: 
```
with XDMFFile(MPI.comm_world, "w.xdmf") as f:
    f.write_checkpoint(v, "v", 0)
```
which works as expected

**Comment by tianyikillua** [comment]
Well here the `dim` for `VectorFunctionSpace` only indicates the number of components and has nothing to do with the spatial / manifold dimension right?

For instance, I often "abusively" store symmetric 2nd-order tensors as 6-dim vectors in 3d, and 4-dim vectors for plane strain cases (`XX`, `YY`, `ZZ`, `XY`).

**Comment by tianyikillua** [comment]
> which works as expected

Yes for `dim <= 3`. Crash for `dim >= 4`.

**Comment by jorgensd** [comment]
> > which works as expected
> 
> Yes for `dim <= 3`. Crash for `dim >= 4`.

I dont know if XDMF support 4D elements.

**Comment by jorgensd** [comment]
@michalhabera , do you have any additional insight here?

**Comment by jorgensd** [comment]
The bug is in this line: https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfin/io/XDMFFile.cpp#L705

I'll have a look at it later today

**Comment by jorgensd** [comment]
@tianyikillua proposed fix in #667 . I tested this locally for dim=2,3,4 with `XDMFFile.write`. Could you verify that this solves your issue?

**Comment by tianyikillua** [comment]
> @tianyikillua proposed fix in #667 . I tested this locally for dim=2,3,4 with `XDMFFile.write`. Could you verify that this solves your issue?

Thanks! I'll try to compile `dolfinx` then...

**Comment by tianyikillua** [comment]
> @tianyikillua proposed fix in #667 . I tested this locally for dim=2,3,4 with `XDMFFile.write`. Could you verify that this solves your issue?

Yes it works fine for me! Thanks again.

**Comment by garth-wells** [comment]
I don't know how to properly fix this. Maybe the user should be required to indicate that the element is DG0.
